VirtoCommerce.Storefront/Domain/Catalog/ExperienceCatalogService.cs

```diff
@@ -119,44 +122,51 @@ public virtual async Task<CatalogSearchResult> SearchProductsAsync(ProductSearch
                 Aggregations = !aggregations.IsNullOrEmpty() ? aggregations.Where(x => aggrIsVisbileSpec.IsSatisfiedBy(x)).ToArray() : new Aggregation[] { }
             };
 
-           
-
             //Post loading initialization of the resulting aggregations
             var aggrContext = new AggregationPostLoadContext
             {
                 ProductSearchCriteria = criteria
             };
+
             var aggrItemCatIds = searchResult.Aggregations.SelectMany(x => x.Items).OfType<CategoryAggregationItem>().Select(x => x.CategoryId).Distinct().ToArray();
             if (aggrItemCatIds.Any())
             {
                 aggrContext.CategoryByIdDict = (await GetCategoriesAsync(aggrItemCatIds, CategoryResponseGroup.Info))
                                                             .Distinct().ToDictionary(x => x.Id)
                                                             .WithDefaultValue(null);
             }
-            searchResult.Aggregations.Apply(x => x.PostLoadInit(aggrContext));
 
+            searchResult.Aggregations.Apply(x => x.PostLoadInit(aggrContext));
 
-            var associationList = searchResult
+            var productAssociations = searchResult
                 .Products
                 .Where(x => x.Associations != null)
                 .SelectMany(x => x.Associations)
                 .ToArray();
 
-            if (!associationList.IsNullOrEmpty())
+            await LoadAssociations(productAssociations);
+
+            return searchResult;
+        }
+
+        public virtual async Task LoadAssociations(ProductAssociation[] productAssociations)
+        {
+            if (productAssociations.IsNullOrEmpty())
             {
-                var allAssociations = await GetProductsAsync(associationList.Select(x => x.ProductId).ToArray());
-                foreach (var association in associationList)
-                {
-                    association.Product = allAssociations.FirstOrDefault(x => x.Id == association.ProductId);
+                return;
+            }
 
-                    if (association.Product != null)
-                    {
-                        EstablishLazyDependenciesForProducts(new[] { association.Product });
-                    }
+            var allAssociations = await GetProductsAsync(productAssociations.Select(x => x.Product.Id).ToArray());
```

Comment:
Why `x.Product.Id` is used instead of `x.ProductId`? Is `ProductId` empty?

Reply:
It contains the wrong id for associations. ProductId is the main id of the product, so we cant get associations with this id.